File: src/wiki/frontmatter_schema.py
"""JSON Schema validation for wiki document frontmatter."""

from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.parse import urlsplit
from urllib.request import Request, urlopen

from jsonschema import Draft202012Validator
from jsonschema.exceptions import ValidationError

from .config import Config
from .graph import _effective_types, resolve_type
from .parser import document_data_from_path
from .paths import iter_document_files, route_for_document_file

logger = logging.getLogger(__name__)

JSON_SCHEMA_KEY = "wazoo:jsonSchema"
TARGET_CLASS_KEY = "sh:targetClass"

REMOTE_FETCH_TIMEOUT = 10
MAX_SCHEMA_BYTES = 1_000_000

SchemaIssue = tuple[str, str]  # (missing_schema_ref | frontmatter_schema, message)


def coerce_schema_refs(value: object) -> list[str] | None:
    """Normalize wazoo:jsonSchema to a non-empty list of strings, or None if absent."""
    if value is None:
        return None
    if isinstance(value, str):
        text = value.strip()
        return [text] if text else None
    if isinstance(value, list):
        refs: list[str] = []
        for item in value:
            if not isinstance(item, str):
                raise ValueError("wazoo:jsonSchema list items must be strings")
            text = item.strip()
            if not text:
                raise ValueError("wazoo:jsonSchema list items must be non-empty strings")
            refs.append(text)
        return refs or None
    raise ValueError("wazoo:jsonSchema must be a string or list of strings")


def is_remote_schema_ref(ref: str) -> bool:
    return ref.startswith("http://") or ref.startswith("https://")


def resolve_local_schema_path(raw: str, config_root: Path) -> Path:
    """Resolve a local wazoo:jsonSchema path relative to the wiki config root."""
    text = raw.strip().replace("\\", "/")
    path = Path(text)
    if not path.is_absolute():
        path = config_root / path
    return path.resolve()


def schema_path_within_root(path: Path, config_root: Path) -> bool:
    try:
        path.resolve().relative_to(config_root.resolve())
    except ValueError:
        return False
    return True


def local_schema_is_valid(path: Path, config_root: Path) -> bool:
    if not schema_path_within_root(path, config_root):
        return False
    return path.is_file() and path.name.lower().endswith(".json")


def normalize_type_uri(type_token: object, config: Config) -> str:
    return str(resolve_type(type_token, config.context))


def is_schema_binding_document(fm_data: dict[str, Any]) -> bool:
    if not fm_data.get(TARGET_CLASS_KEY):
        return False
    try:
        return coerce_schema_refs(fm_data.get(JSON_SCHEMA_KEY)) is not None
    except ValueError:
        return False


def validation_payload(fm_data: dict[str, Any]) -> dict[str, Any]:
    """Frontmatter dict for JSON Schema validation (exclude schema pointers and SHACL keys)."""
    payload: dict[str, Any] = {}
    for key, value in fm_data.items():
        if key.startswith("@"):
            continue
        if key in {"id", JSON_SCHEMA_KEY}:
            continue
        if key.startswith("sh:") or key == TARGET_CLASS_KEY:
            continue
        payload[key] = value
    return payload


def _dedupe_refs(refs: list[str]) -> list[str]:
    seen: set[str] = set()
    ordered: list[str] = []
    for ref in refs:
        if ref in seen:
            continue
        seen.add(ref)
        ordered.append(ref)
    return ordered


def build_type_schema_registry(config: Config) -> dict[str, list[str]]:
    """Map normalized target-class URI to schema reference strings from binding documents."""
    registry: dict[str, list[str]] = {}
    for file_path in iter_document_files(config):
        fm_data = document_data_from_path(file_path, content_predicate=config.graph.content_predicate)
        if not fm_data or not is_schema_binding_document(fm_data):
            continue
        try:
            refs = coerce_schema_refs(fm_data.get(JSON_SCHEMA_KEY))
        except ValueError:
            continue
        if not refs:
            continue
        target = fm_data.get(TARGET_CLASS_KEY)
        if target is None:
            continue
        type_key = normalize_type_uri(target, config)
        registry.setdefault(type_key, []).extend(refs)
    return {key: _dedupe_refs(refs) for key, refs in registry.items()}


class SchemaLoader:
    """Load and cache JSON Schema documents from local paths or remote URLs."""

    def __init__(
        self,
        config_root: Path,
        *,
        remote_schema_refs: str = "allow",
        remote_schema_hosts: list[str] | None = None,
    ) -> None:
        self.config_root = config_root.resolve()
        self.remote_schema_refs = remote_schema_refs
        self.remote_schema_hosts = set(remote_schema_hosts or [])
        self._schema_cache: dict[str, tuple[dict[str, Any] | None, str | None]] = {}
        self._validator_cache: dict[str, tuple[Draft202012Validator | None, str | None]] = {}

    def _remote_policy_error(self, ref: str) -> str | None:
        if self.remote_schema_refs == "allow":
            return None
        if self.remote_schema_refs == "deny":
            return "remote schema refs are disabled by check.remote_schema_refs"
        host = urlsplit(ref).hostname
        if host is None or host not in self.remote_schema_hosts:
            label = host or ref
            return f"remote schema host {label!r} is not allowed by check.remote_schema_hosts"
        return None

    def load_schema(self, ref: str) -> tuple[dict[str, Any] | None, str | None]:
        if ref in self._schema_cache:
            return self._schema_cache[ref]

        if is_remote_schema_ref(ref):
            policy_error = self._remote_policy_error(ref)
            if policy_error:
                schema, error = None, policy_error
            else:
                schema, error = self._fetch_remote(ref)
        else:
            schema, error = self._load_local(ref)

        self._schema_cache[ref] = (schema, error)
        return schema, error

    def get_validator(self, ref: str) -> tuple[Draft202012Validator | None, str | None]:
        if ref in self._validator_cache:
            return self._validator_cache[ref]

        schema, error = self.load_schema(ref)
        if error or schema is None:
            self._validator_cache[ref] = (None, error)
            return None, error

        try:
            validator = Draft202012Validator(schema)
        except Exception as exc:
            message = f"invalid JSON Schema document ({exc})"
            self._validator_cache[ref] = (None, message)
            return None, message

        self._validator_cache[ref] = (validator, None)
        return validator, None

    def _load_local(self, ref: str) -> tuple[dict[str, Any] | None, str | None]:
        path = resolve_local_schema_path(ref, self.config_root)
        if not local_schema_is_valid(path, self.config_root):
            return None, "must resolve to a readable .json file under the wiki config root"
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError) as exc:
            return None, f"could not be read as JSON ({exc})"
        if not isinstance(data, dict):
            return None, "must contain a JSON object"
        return data, None

    def _fetch_remote(self, ref: str) -> tuple[dict[str, Any] | None, str | None]:
        request = Request(ref, headers={"User-Agent": "Wiki-CLI/jsonschema"})
        try:
            with urlopen(request, timeout=REMOTE_FETCH_TIMEOUT) as response:
                raw = response.read(MAX_SCHEMA_BYTES + 1)
        except HTTPError as exc:
            return None, f"HTTP {exc.code}"
        except URLError as exc:
            return None, str(exc.reason)
        except TimeoutError:
            return None, "request timed out"
        except OSError as exc:
            return None, str(exc)

        if len(raw) > MAX_SCHEMA_BYTES:
            return None, f"response exceeds {MAX_SCHEMA_BYTES} bytes"

        try:
            data = json.loads(raw.decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError) as exc:
            return None, f"response is not valid JSON ({exc})"
        if not isinstance(data, dict):
            return None, "response must be a JSON object"
        return data, None


def _effective_schema_refs(
    fm_data: dict[str, Any],
    registry: dict[str, list[str]],
    config: Config,
) -> list[tuple[str, str | None]]:
    """Return (schema_ref, via_label) pairs to validate, deduped."""
    refs: list[tuple[str, str | None]] = []
    seen: set[str] = set()

    for type_token in _effective_types(fm_data, config):
        type_uri = normalize_type_uri(type_token, config)
        for ref in registry.get(type_uri, []):
            if ref in seen:
                continue
            seen.add(ref)
            refs.append((ref, f"via type {type_token}"))

    if not is_schema_binding_document(fm_data):
        try:
            page_refs = coerce_schema_refs(fm_data.get(JSON_SCHEMA_KEY))
        except ValueError:
            page_refs = None
        if page_refs:
            for ref in page_refs:
                if ref in seen:
                    continue
                seen.add(ref)
                refs.append((ref, None))

    return refs


def _format_missing_ref(route: str, ref: str, detail: str, *, binding: bool) -> str:
    if is_remote_schema_ref(ref):
        return f"In {route}: wazoo:jsonSchema {ref!r} could not be fetched ({detail})."
    if binding:
        return (
            f"In {route}: wazoo:jsonSchema on type binding {ref!r} {detail}."
        )
    return f"In {route}: wazoo:jsonSchema {ref!r} {detail}."


def _format_validation_error(route: str, ref: str, error: ValidationError, via: str | None) -> str:
    via_part = f", {via}" if via else ""
    return f"In {route}: {error.message} (schema: {ref}{via_part})"


def check_frontmatter_schema(
    config: Config,
    file_filter: set[str] | None = None,
    *,
    file_paths: list[Path] | None = None,
) -> tuple[list[str], list[str]]:
    """Return (missing_schema_ref issues, frontmatter_schema validation issues)."""
    if config.check.frontmatter_schema == "off" and config.check.missing_schema_ref == "off":
        return [], []

    registry = build_type_schema_registry(config)
    loader = SchemaLoader(
        config.config_root,
        remote_schema_refs=config.check.remote_schema_refs,
        remote_schema_hosts=config.check.remote_schema_hosts,
    )
    missing_issues: list[str] = []
    validation_issues: list[str] = []

    candidates = file_paths if file_paths is not None else iter_document_files(config)

    for file_path in candidates:
        try:
            route = route_for_document_file(config, file_path)
        except ValueError:
            continue
        if file_filter is not None and route not in file_filter:
            continue

        fm_data = document_data_from_path(file_path, content_predicate=config.graph.content_predicate)
        if not fm_data:
            continue

        try:
            coerce_schema_refs(fm_data.get(JSON_SCHEMA_KEY))
        except ValueError as exc:
            validation_issues.append(f"In {route}: {exc}.")
            continue

        if is_schema_binding_document(fm_data):
            try:
                binding_refs = coerce_schema_refs(fm_data.get(JSON_SCHEMA_KEY)) or []
            except ValueError:
                binding_refs = []
            for ref in binding_refs:
                _, error = loader.load_schema(ref)
                if error and config.check.missing_schema_ref != "off":
                    missing_issues.append(
                        _format_missing_ref(route, ref, error, binding=True)
                    )
            continue

        schema_refs = _effective_schema_refs(fm_data, registry, config)
        if not schema_refs:
            continue

        instance = validation_payload(fm_data)
        for ref, via in schema_refs:
            validator, error = loader.get_validator(ref)
            if error:
                if config.check.missing_schema_ref != "off":
                    missing_issues.append(_format_missing_ref(route, ref, error, binding=False))
                continue
            if validator is None:
                continue
            for validation_error in sorted(validator.iter_errors(instance), key=lambda e: list(e.path)):
                if config.check.frontmatter_schema != "off":
                    validation_issues.append(
                        _format_validation_error(route, ref, validation_error, via)
                    )

    return missing_issues, validation_issues

File: src/wiki/frontmatter_schema.ts
/**
 * JSON Schema validation for wiki document frontmatter.
 *
 * Port of `src/wiki/frontmatter_schema.py`. Pages declare a schema two ways —
 * a *binding document* carries `sh:targetClass` plus `wazoo:jsonSchema` and
 * applies that schema to every page of the class, and a page may name extra
 * schemas for itself — and this module turns those declarations into the two
 * issue lists `wiki check` reports under `missing_schema_ref` and
 * `frontmatter_schema`.
 *
 * The validation language itself is {@link JsonSchemaValidator}, not this file;
 * what lives here is the lookup, the caching, the remote-fetch policy, and the
 * message shapes.
 *
 * Two deliberate differences from the Python module:
 *
 * - **The entry point is `async`.** Python resolves a remote schema with a
 *   blocking `urlopen`; the port uses `fetch`. The split between "reported
 *   issue" and "value" is unchanged, only the `await` is new, and every local
 *   path still performs no I/O beyond reading the schema file.
 * - **A malformed JSON schema file reports a JavaScript parse message.** Python
 *   renders `json.JSONDecodeError` (`Expecting value: line 1 column 1 (char 0)`);
 *   `JSON.parse` renders its own prose. The `could not be read as JSON (...)`
 *   frame, the issue's classification, and the exit code are identical, so this
 *   is spec-close by the ADR's rule rather than a bug to chase.
 *
 * Remote fetching is likewise environment-dependent by nature (DNS failures,
 * TLS errors) and is never byte-compared; the tests exercise it through an
 * injected fetch, the way the Python tests patch `urlopen`.
 */

import type { Config } from "./config.ts";
import { Path, ValueError } from "./fspath.ts";
import { effectiveTypes, resolveType } from "./graph.ts";
import { JsonSchemaValidator, sortByInstancePath } from "./json_schema.ts";
import {
  type DataRecord,
  documentDataFromPath,
  isRecord,
  pyStrip,
  readTextTolerant,
} from "./parser.ts";
import { iterDocumentFiles, routeForDocumentFile } from "./paths.ts";
import { pyStr } from "./pyrepr.ts";

/** The frontmatter key naming a JSON Schema document. */
export const JSON_SCHEMA_KEY = "wazoo:jsonSchema";

/** The SHACL key that turns a document into a type binding. */
export const TARGET_CLASS_KEY = "sh:targetClass";

/** Seconds before a remote schema fetch is abandoned. */
export const REMOTE_FETCH_TIMEOUT = 10;

/** Largest remote schema document the engine will read. */
export const MAX_SCHEMA_BYTES = 1_000_000;

/** The two issue lists `check_frontmatter_schema` returns. */
export type SchemaIssues = readonly [string[], string[]];

/**
 * Normalize `wazoo:jsonSchema` to a non-empty list of strings, or `null` when
 * absent.
 *
 * A scalar becomes a one-element list, whitespace-only entries are dropped, and
 * anything that is not a string raises the same `ValueError` the Python side
 * raises — the caller turns that into an `In <route>: ...` issue with the
 * exception's own text.
 */
export function coerceSchemaRefs(value: unknown): string[] | null {
  if (value === null || value === undefined) return null;
  if (typeof value === "string") {
    const text = pyStrip(value);
    return text === "" ? null : [text];
  }
  if (Array.isArray(value)) {
    const refs: string[] = [];
    for (const item of value) {
      if (typeof item !== "string") {
        throw new ValueError(
          "wazoo:jsonSchema list items must be strings",
        );
      }
      const text = pyStrip(item);
      if (text === "") {
        throw new ValueError(
          "wazoo:jsonSchema list items must be non-empty strings",
        );
      }
      refs.push(text);
    }
    return refs.length === 0 ? null : refs;
  }
  throw new ValueError("wazoo:jsonSchema must be a string or list of strings");
}

/** `true` for the two schemes the engine will fetch. */
export function isRemoteSchemaRef(ref: string): boolean {
  return ref.startsWith("http://") || ref.startsWith("https://");
}

/** Resolve a local `wazoo:jsonSchema` path relative to the wiki config root. */
export function resolveLocalSchemaPath(raw: string, configRoot: Path): Path {
  const text = pyStrip(raw).replaceAll("\\", "/");
  const path = Path.of(text);
  return (path.isAbsolute() ? path : configRoot.joinpath(path)).resolve();
}

/** `true` when `path` lives under `configRoot`. */
export function schemaPathWithinRoot(path: Path, configRoot: Path): boolean {
  try {
    path.resolve().relativeTo(configRoot.resolve());
  } catch (error) {
    if (error instanceof ValueError) return false;
    throw error;
  }
  return true;
}

/** `true` when `path` is a readable `.json` file under the config root. */
export function localSchemaIsValid(path: Path, configRoot: Path): boolean {
  if (!schemaPathWithinRoot(path, configRoot)) return false;
  return path.isFile() && path.name.toLowerCase().endsWith(".json");
}

/**
 * The type URI a `sh:targetClass` value names.
 *
 * `resolve_type` is a CURIE expander, not a string cast, so `schema:Person`
 * and the bare `Person` land on different URIs by design. The Python side
 * stringifies the result — `None` becomes the string `"None"` — and the
 * registry's keys depend on that, so the port keeps `String(None)`.
 */
export function normalizeTypeUri(typeToken: unknown, config: Config): string {
  const resolved = resolveType(typeToken, config.context);
  return resolved === null ? "None" : resolved.value;
}

/** `true` when a document binds a schema to a type class. */
export function isSchemaBindingDocument(fmData: DataRecord): boolean {
  if (!pyTruthy(fmData[TARGET_CLASS_KEY])) return false;
  try {
    return coerceSchemaRefs(fmData[JSON_SCHEMA_KEY]) !== null;
  } catch (error) {
    if (error instanceof ValueError) return false;
    throw error;
  }
}

/**
 * The frontmatter a JSON Schema validates.
 *
 * Schema pointers and SHACL keys are metadata about the binding, not content of
 * the page, so validating them against the page schema would report failures
 * the author cannot fix without deleting the binding.
 */
export function validationPayload(
  fmData: DataRecord,
): Record<string, unknown> {
  const payload: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(fmData)) {
    if (key.startsWith("@")) continue;
    if (key === "id" || key === JSON_SCHEMA_KEY) continue;
    if (key.startsWith("sh:") || key === TARGET_CLASS_KEY) continue;
    payload[key] = value;
  }
  return payload;
}

/** Keep the first occurrence of each reference, in order. */
function dedupeRefs(refs: readonly string[]): string[] {
  const seen = new Set<string>();
  const ordered: string[] = [];
  for (const ref of refs) {
    if (seen.has(ref)) continue;
    seen.add(ref);
    ordered.push(ref);
  }
  return ordered;
}

/**
 * Map a normalized target-class URI to the schema references binding it.
 *
 * Every document is read, not only the ones being checked: a page is validated
 * against a schema declared in *another* page, so `check --path` still needs
 * the whole registry.
 */
export function buildTypeSchemaRegistry(
  config: Config,
): Map<string, string[]> {
  const registry = new Map<string, string[]>();
  for (const filePath of iterDocumentFiles(config)) {
    const fmData = documentDataFromPath(
      filePath,
      config.graph.content_predicate ?? undefined,
    );
    if (fmData === null || !isSchemaBindingDocument(fmData)) continue;
    let refs: string[] | null;
    try {
      refs = coerceSchemaRefs(fmData[JSON_SCHEMA_KEY]);
    } catch (error) {
      if (error instanceof ValueError) continue;
      throw error;
    }
    if (refs === null || refs.length === 0) continue;
    const target = fmData[TARGET_CLASS_KEY];
    if (target === null || target === undefined) continue;
    const typeKey = normalizeTypeUri(target, config);
    const existing = registry.get(typeKey);
    if (existing === undefined) registry.set(typeKey, [...refs]);
    else existing.push(...refs);
  }
  for (const [key, refs] of registry) registry.set(key, dedupeRefs(refs));
  return registry;
}

/** How an unexpected property name reaches the `additionalProperties` message. */
type Fetch = typeof fetch;

export interface SchemaLoaderOptions {
  readonly remoteSchemaRefs?: "allow" | "deny" | "allowlist";
  readonly remoteSchemaHosts?: readonly string[];
  /** Injected for tests, the way the Python tests patch `urlopen`. */
  readonly fetch?: Fetch;
}

/** A loaded schema, or the reason it could not be loaded. */
interface LoadedSchema {
  readonly schema: Record<string, unknown> | null;
  readonly error: string | null;
}

/** A compiled validator, or the reason there is none. */
interface LoadedValidator {
  readonly validator: JsonSchemaValidator | null;
  readonly error: string | null;
}

/** Load and cache JSON Schema documents from local paths or remote URLs. */
export class SchemaLoader {
  readonly configRoot: Path;
  readonly remoteSchemaRefs: "allow" | "deny" | "allowlist";
  readonly remoteSchemaHosts: ReadonlySet<string>;
  readonly #fetch: Fetch;
  readonly #schemaCache = new Map<string, LoadedSchema>();
  readonly #validatorCache = new Map<string, LoadedValidator>();

  constructor(configRoot: Path, options: SchemaLoaderOptions = {}) {
    this.configRoot = configRoot.resolve();
    this.remoteSchemaRefs = options.remoteSchemaRefs ?? "allow";
    this.remoteSchemaHosts = new Set(options.remoteSchemaHosts ?? []);
    this.#fetch = options.fetch ?? fetch;
  }

  /**
   * The configured policy's objection to a remote reference, if any.
   *
   * Checked *before* any network call, so `deny` and an unmatched `allowlist`
   * host produce an issue without a request — which is the whole point of the
   * setting.
   */
  #remotePolicyError(ref: string): string | null {
    if (this.remoteSchemaRefs === "allow") return null;
    if (this.remoteSchemaRefs === "deny") {
      return "remote schema refs are disabled by check.remote_schema_refs";
    }
    const host = urlHostname(ref);
    if (host === null || !this.remoteSchemaHosts.has(host)) {
      return `remote schema host ${
        pyReprStr(host ?? ref)
      } is not allowed by check.remote_schema_hosts`;
    }
    return null;
  }

  /** Load a schema document, or report why it could not be loaded. */
  async loadSchema(ref: string): Promise<LoadedSchema> {
    const cached = this.#schemaCache.get(ref);
    if (cached !== undefined) return cached;

    let result: LoadedSchema;
    if (isRemoteSchemaRef(ref)) {
      const policyError = this.#remotePolicyError(ref);
      result = policyError !== null
        ? { schema: null, error: policyError }
        : await this.#fetchRemote(ref);
    } else {
      result = this.#loadLocal(ref);
    }

    this.#schemaCache.set(ref, result);
    return result;
  }

  /** Compile a schema document once, caching both outcomes. */
  async getValidator(ref: string): Promise<LoadedValidator> {
    const cached = this.#validatorCache.get(ref);
    if (cached !== undefined) return cached;

    const { schema, error } = await this.loadSchema(ref);
    if (error !== null || schema === null) {
      const result = { validator: null, error };
      this.#validatorCache.set(ref, result);
      return result;
    }

    let result: LoadedValidator;
    try {
      result = { validator: new JsonSchemaValidator(schema), error: null };
    } catch (compileError) {
      // Both construction failures and mid-validation schema errors reach
      // here: the port compiles where jsonschema validated lazily, which turns
      // a traceback into a reported issue (see json_schema.ts).
      result = {
        validator: null,
        error: `invalid JSON Schema document (${
          compileError instanceof Error
            ? compileError.message
            : String(compileError)
        })`,
      };
    }
    this.#validatorCache.set(ref, result);
    return result;
  }

  #loadLocal(ref: string): LoadedSchema {
    const path = resolveLocalSchemaPath(ref, this.configRoot);
    if (!localSchemaIsValid(path, this.configRoot)) {
      return {
        schema: null,
        error:
          "must resolve to a readable .json file under the wiki config root",
      };
    }
    let data: unknown;
    try {
      data = JSON.parse(readTextTolerant(path)) as unknown;
    } catch (error) {
      return { schema: null, error: `could not be read as JSON (${error})` };
    }
    if (!isRecord(data)) {
      return { schema: null, error: "must contain a JSON object" };
    }
    return { schema: data, error: null };
  }

  async #fetchRemote(ref: string): Promise<LoadedSchema> {
    let response: Response;
    try {
      response = await this.#fetch(ref, {
        headers: { "User-Agent": "Wiki-CLI/jsonschema" },
        signal: AbortSignal.timeout(REMOTE_FETCH_TIMEOUT * 1000),
      });
    } catch (error) {
      return { schema: null, error: fetchFailureMessage(error) };
    }

    if (!response.ok) {
      // `urlopen` raises on a non-2xx status; `fetch` resolves, so the
      // equivalent verdict is spelled out here.
      await response.body?.cancel().catch(() => {});
      return { schema: null, error: `HTTP ${response.status}` };
    }

    let raw: Uint8Array;
    try {
      raw = new Uint8Array(await response.arrayBuffer());
    } catch (error) {
      return { schema: null, error: fetchFailureMessage(error) };
    }
    if (raw.byteLength > MAX_SCHEMA_BYTES) {
      return {
        schema: null,
        error: `response exceeds ${MAX_SCHEMA_BYTES} bytes`,
      };
    }

    let data: unknown;
    try {
      data = JSON.parse(new TextDecoder().decode(raw)) as unknown;
    } catch (error) {
      return { schema: null, error: `response is not valid JSON (${error})` };
    }
    if (!isRecord(data)) {
      return { schema: null, error: "response must be a JSON object" };
    }
    return { schema: data, error: null };
  }
}

/**
 * A one-line reason for a failed fetch.
 *
 * Python renders `URLError.reason` (`[Errno 11001] getaddrinfo failed`);
 * `fetch` rejects with a `TypeError` whose `cause` carries the platform error.
 * The cause is the closest available text and, unlike the Python spelling, it
 * is not stable across runtimes — which is why the parity harness never
 * compares a live network failure.
 */
function fetchFailureMessage(error: unknown): string {
  if (error instanceof DOMException && error.name === "TimeoutError") {
    return "request timed out";
  }
  const cause = (error as { cause?: unknown }).cause;
  const source = cause ?? error;
  return source instanceof Error ? source.message : String(source);
}

/** `urllib.parse.urlsplit(ref).hostname`, which is lower-cased and port-free. */
function urlHostname(ref: string): string | null {
  try {
    const url = new URL(ref);
    return url.hostname === "" ? null : url.hostname;
  } catch {
    return null;
  }
}

/** `!value`, in Python's sense, for the falsy checks this module inherits. */
function pyTruthy(value: unknown): boolean {
  if (value === null || value === undefined) return false;
  if (typeof value === "boolean") return value;
  if (typeof value === "number") return value !== 0;
  if (typeof value === "string") return value !== "";
  if (Array.isArray(value)) return value.length > 0;
  if (isRecord(value)) return Object.keys(value).length > 0;
  return true;
}

/** Python's `repr` for a string, used by the host-policy message. */
function pyReprStr(text: string): string {
  return text.includes("'") && !text.includes('"') ? `"${text}"` : `'${text}'`;
}

/**
 * The schema references that apply to one document.
 *
 * Type bindings come first, then the page's own `wazoo:jsonSchema`, each
 * deduped across both sources. A page that *is* a binding document contributes
 * no page-level refs of its own: its `wazoo:jsonSchema` describes the class, so
 * validating the binding page against it would fail on the binding's own keys.
 */
function effectiveSchemaRefs(
  fmData: DataRecord,
  registry: ReadonlyMap<string, string[]>,
  config: Config,
): [string, string | null][] {
  const refs: [string, string | null][] = [];
  const seen = new Set<string>();

  for (const typeToken of effectiveTypes(fmData, config)) {
    const typeUri = normalizeTypeUri(typeToken, config);
    for (const ref of registry.get(typeUri) ?? []) {
      if (seen.has(ref)) continue;
      seen.add(ref);
      refs.push([ref, `via type ${pyStr(typeToken)}`]);
    }
  }

  if (!isSchemaBindingDocument(fmData)) {
    let pageRefs: string[] | null = null;
    try {
      pageRefs = coerceSchemaRefs(fmData[JSON_SCHEMA_KEY]);
    } catch (error) {
      if (!(error instanceof ValueError)) throw error;
    }
    for (const ref of pageRefs ?? []) {
      if (seen.has(ref)) continue;
      seen.add(ref);
      refs.push([ref, null]);
    }
  }

  return refs;
}

/** The issue for a `wazoo:jsonSchema` that could not be loaded at all. */
function formatMissingRef(
  route: string,
  ref: string,
  detail: string,
  options: { readonly binding: boolean },
): string {
  if (isRemoteSchemaRef(ref)) {
    return `In ${route}: wazoo:jsonSchema ${
      pyReprStr(ref)
    } could not be fetched (${detail}).`;
  }
  if (options.binding) {
    return `In ${route}: wazoo:jsonSchema on type binding ${
      pyReprStr(ref)
    } ${detail}.`;
  }
  return `In ${route}: wazoo:jsonSchema ${pyReprStr(ref)} ${detail}.`;
}

/** The issue for one schema validation failure. */
function formatValidationError(
  route: string,
  ref: string,
  error: { readonly message: string },
  via: string | null,
): string {
  const viaPart = via ? `, ${via}` : "";
  return `In ${route}: ${error.message} (schema: ${ref}${viaPart})`;
}

/**
 * Check every document's frontmatter against the schemas that bind it.
 *
 * Returns `[missing_schema_ref issues, frontmatter_schema issues]`: failures to
 * *find* a schema and failures to *satisfy* one are different rules with
 * different severities, so the caller routes them separately. Both rules must be
 * off before the whole pass — registry build included — is skipped.
 */
export async function checkFrontmatterSchema(
  config: Config,
  fileFilter: ReadonlySet<string> | null = null,
  options: { readonly filePaths?: readonly Path[] | null } = {},
): Promise<SchemaIssues> {
  if (
    config.check.frontmatter_schema === "off" &&
    config.check.missing_schema_ref === "off"
  ) {
    return [[], []];
  }

  const registry = buildTypeSchemaRegistry(config);
  const loader = new SchemaLoader(config.config_root, {
    remoteSchemaRefs: config.check.remote_schema_refs,
    remoteSchemaHosts: config.check.remote_schema_hosts,
  });
  const missingIssues: string[] = [];
  const validationIssues: string[] = [];

  const candidates = options.filePaths ?? iterDocumentFiles(config);

  for (const filePath of candidates) {
    let route: string;
    try {
      route = routeForDocumentFile(config, filePath);
    } catch (error) {
      if (error instanceof ValueError) continue;
      throw error;
    }
    if (fileFilter !== null && !fileFilter.has(route)) continue;

    const fmData = documentDataFromPath(
      filePath,
      config.graph.content_predicate ?? undefined,
    );
    if (fmData === null || !pyTruthy(fmData)) continue;

    try {
      coerceSchemaRefs(fmData[JSON_SCHEMA_KEY]);
    } catch (error) {
      if (!(error instanceof ValueError)) throw error;
      validationIssues.push(`In ${route}: ${error.message}.`);
      continue;
    }

    if (isSchemaBindingDocument(fmData)) {
      let bindingRefs: string[] | null = null;
      try {
        bindingRefs = coerceSchemaRefs(fmData[JSON_SCHEMA_KEY]);
      } catch {
        bindingRefs = null;
      }
      for (const ref of bindingRefs ?? []) {
        const { error } = await loader.loadSchema(ref);
        if (error !== null && config.check.missing_schema_ref !== "off") {
          missingIssues.push(
            formatMissingRef(route, ref, error, { binding: true }),
          );
        }
      }
      continue;
    }

    const schemaRefs = effectiveSchemaRefs(fmData, registry, config);
    if (schemaRefs.length === 0) continue;

    const instance = validationPayload(fmData);
    for (const [ref, via] of schemaRefs) {
      const { validator, error } = await loader.getValidator(ref);
      if (error !== null) {
        if (config.check.missing_schema_ref !== "off") {
          missingIssues.push(
            formatMissingRef(route, ref, error, { binding: false }),
          );
        }
        continue;
      }
      if (validator === null) continue;
      for (const failure of sortByInstancePath(validator.errors(instance))) {
        if (config.check.frontmatter_schema === "off") continue;
        validationIssues.push(
          formatValidationError(route, ref, failure, via),
        );
      }
    }
  }

  return [missingIssues, validationIssues];
}
